add swagger doc using utoipa resolves #85 (#101)
ElektraInitiative/PermaplanT: doc/decisions/backend_api_documentation.md

# Backend API documentation

## Problem

We need a way to document the API in the backend. Otherwise we would have to read the Rust Code every time we want to develop a new part of the frontend.

## Constraints

## Assumptions

> 

Reviewer: ```suggestion

- the backend's API is only used by the frontend

```

## Solutions

### Utoipa

You can think of [Utoipa](https://github.com/juhaku/utoipa) as the Rust equivalent to Swagger/OpenApi in Java.

It is a Rust crate which allows you to annotate your code (structs and endpoints) and automatically generate OpenApi documentation from that. You can also generate a UI endpoint just like with Java Spring.

Compared to swagger in Java there is a bit more you have to configure to get it running.

Once you get it running you get endpoints with schemas and examples representing the API. You can also execute example requests directly from the UI to see how the backend behaves.

### API Blueprint

Information from <https://testfully.io/blog/api-blueprint/>.

Before starting on a new endpoint we would write a markdown document specifying in detail what the endpoint should return and in what format. There is also a [lot of tooling](https://apiblueprint.org/tools.html) for API Blueprint.

> 

Reviewer: Please write also about other alternatives you found. Some investigations should be done.

Author: Do you mean alternatives in the sense of crates/mechanisms similar to Utoipa or API blueprint? Or rather investigate API blueprint in more detail?

Reviewer: I meant other crates but you already said it in the meeting: it seems like there is not really something else available.

## Decision

> Utoipa

Reviewer: Write a bit how exactly we want to use it.

## Rationale

With Utoipa it is possible to generate API documentation directly from code whereas with API Blueprint we would have to write the API documentation by hand in markdown files (at least I have yet to find a crate that automatically generates required markdown files). Additionally if we where to use API Blueprint every time the API changes we would have to remember to change the markdown files as well, which is easily forgotten and might become tedious when making smaller changes.

For me (@kitzbergerg) it is easier to use Utoipa as it generates Swagger/OpenApi which I have already used. I would have to learn API Blueprint and its tooling from scratch. With that I would argue that more people know Swagger/OpenApi leading to easier onboarding.

One advantage of API Blueprint would be that we have to carefully think about the API (API-First Approach) before actually implementing. This leads to a clean API in the long run, however it is more difficult to quickly prototype and try different approaches (which in my opinion is currently more important as we are somewhat at the beginning of the project).

> 

Reviewer: Please write about which data types are supported by Utoipa, e.g., are SQL or Diesel Types supported, e.g. https://lib.rs/crates/postgis_diesel? How much time would it need to add support for them?

Author: I don't quite understand this question. As far as I can tell all basic Rust types are supported as well as some a bit more special ones: https://github.com/juhaku/utoipa#crate-features. We also only have to use Utoipa for the DTOs, the types used in the database should not be exposed in the API anyways.

Apart from that there is also the option to define `#[schema(value_type = String, format = DateTime)]` so that another type or some format is used in the API doc.

## Implications

> -   When deciding on a new endpoint it will take longer for the API documentation to be done, as the backend code has to be finished first.

Author: What do you mean by that?
You have to derive ToSchema and you also might need to add something to the API config. But do we want implementation details like that in the decision?

> -   When deciding on a new endpoint it will take longer for the API documentation to be done, as the backend code has to be finished first.

Reviewer: In this case it seem to be relevant . At least I would have understood with "backend code has to be finished" nearly the opposite to "You have to derive ToSchema and you also might need to add something to the API config".

```suggestion
-   When deciding on a new endpoint it will take longer for the API documentation to be done, as:
    - the Rust structs must be implemented
    - `derive ToSchema` must be present for all structs, and
    - you also might need to change the API config.
```

-   The API documentation will not deviate from the actual implementation as it is generated from code.
-   Developers have to know Swagger, but once they do they can easily check the API in the browser served by the backend.
-   Backend Developers have to invest some time into reading Utoipa's documentation.

## Related Decisions

## Notes

> Once the project is in a later stage we might consider additionally using API Blueprint. At that point a cleaner API becomes more important and making changes requires more careful consideration.

Reviewer: Not needed to be written. If assumptions are broken, the decision needs to be reevaluated again.

```suggestion
```
